### ISSUE TITLE
Compiler should warn about allocated but unused qubits

### ISSUE THREAD

**Issue by tcNickolas** [microsoft/qsharp-compiler#584]
**Is your feature request related to a problem? Please describe.**
When we define a classical variable that is never used later, the compiler offers a warning, which is very helpful. However, when we allocate a qubit register that is not used later, no warning is offered. 

As a repro, this code only gives warning for variable `a` (`warning CS0219: The variable 'a' is assigned but its value is never used`), not `q`:
```
using (q = Qubit()) {
    let a = 1;
}
```

**Describe the solution you'd like**
It would be really nice to offer a warning for unused qubits, especially now that a lot of Q# code is executed on a simulator, and allocating extra qubits makes the simulation slower.

**Context**
QDK 0.12.20072031.

**Comment by samarsha** [comment]
Yes, I agree, this would be useful. Actually, I think the problem is that the Q# compiler does not produce any warnings for unused variables. The warning you see for classical variables is from the C# compiler. In the generated C#, all qubit variables are released after they are allocated, even if they are never used, and the release is considered a use according to the C# compiler. So no warning is emitted for qubits.

I think it makes sense for consistency to add warnings for all unused variables (lets, mutables, usings, maybe callable parameters?) to the Q# compiler, so we aren't relying on the C# compiler for any of them.

**Comment by cgranade** [comment]
> I think it makes sense for consistency to add warnings for all unused variables (lets, mutables, usings, maybe callable parameters?) to the Q# compiler, so we aren't relying on the C# compiler for any of them.

At least with respect to callable parameters, there can be very valid reasons to not using some parameters (the `delay : Unit` input to `Delay` comes to mind, for instance), such that I think it'd make sense to have a mechanism for disabling that warning. Perhaps a `@NoWarn` attribute?

**Comment by samarsha** [comment]
It looks like `_` is currently not allowed for parameter names, like it is for `let` bindings, but what about disabling the warning for parameters named `_`?

**Comment by tcNickolas** [comment]
I would love to be able to disable any warning! The one about possible shortcuts in ternary expressions is quite persistent...

**Comment by samarsha** [comment]
The Q# compiler does have the `--no-warn` option, but it looks like there's no easy way to set that from the project config or a Q# file...
